test/es-module/test-vm-contextified-script-leak.js

```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+// Flags: --max-old-space-size=16 --trace-gc
+
+require('../common');
+const vm = require('vm');
+let count = 0;
+
+function main() {
+  new vm.Script(`"${Math.random().toString().repeat(512)}";`, {
```

Comment:
I dug into the memory layout a bit and I found out that by fixing the use-after-free and making sure the `ContextifyScript` don't get freed as long as the corresponding `vm.Script` is still alive, we now have a leak caused by the persistent handle `v8::UnboundScript`, which is similar to the issue caused by the persistent handle to `v8::Module` in `vm.SourceTextModule`, and is what this test can reproduce.

One way to fix all this is to retain a reference to some `v8::Data` in a way that V8's GC understands. To my knowledge there's no existing V8 API that allows this (only `v8::Value` is supported so far). I added a pre-mature V8 patch here for the POC, together with the last commit here locally it fixes the leak for me. Without a V8 API like that I think we'd probably have to choose between a use-after-free or a leak, so I'll try to get a solution in the upstream.

Reply:
Opened issue in the upstream https://bugs.chromium.org/p/v8/issues/detail?id=14120

Reply:
I have a WIP patch here https://chromium-review.googlesource.com/c/v8/v8/+/4707972 not sure if the whole idea is acceptable though. Will send it for review later.

Reply:
The V8 patch for the `SetInternalField()` change has landed in the upstream, backport opened in https://github.com/nodejs/node/pull/49419

This also needs another smaller ABI breaking patch to work though, opened in the upstream at : https://chromium-review.googlesource.com/c/v8/v8/+/4827307

Reply:
The required V8 patches have already been backported to the main branch.